File: Java/construct-binary-tree-from-inorder-and-postorder-traversal.java
//  Construct Binary Tree from Inorder and Postorder Traversal

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// Time O(N), Space O(N)
class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if (inorder.length!=postorder.length || inorder.length==0)
            return null;
        int size = inorder.length, root_val = postorder[size-1];
        int left_size = 0, right_size = 0;
        for (int i=0; i<size && inorder[i]!=root_val; i++)
            left_size ++;
        right_size = size - 1 - left_size;
        int[] left_inorder = new int[left_size];
        int[] left_postorder = new int[left_size];
        int[] right_inorder = new int[right_size];
        int[] right_postorder = new int[right_size];
        for (int i=0; i<left_size; i++) {
            left_inorder[i] = inorder[i];
            left_postorder[i] = postorder[i];
        }
        for (int i=0; i<right_size; i++) {
            right_inorder[i] = inorder[i+left_size+1];
            right_postorder[i] = postorder[i+left_size];
        }
        TreeNode root = new TreeNode(root_val);
        root.left = buildTree(left_inorder, left_postorder);
        root.right = buildTree(right_inorder, right_postorder);
        return root;
    }
}

File: Leetcode150/construct-binary-tree-from-inorder-and-postorder-traversal.cc
//  Construct Binary Tree from Inorder and Postorder Traversal

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        return buildTreeWithSegment(inorder, 0, inorder.size()-1,
                                   postorder, 0, postorder.size()-1);
    }
    
    TreeNode *buildTreeWithSegment(vector<int> &inorder, int l1, int r1,
                                  vector<int> &postorder, int l2, int r2) {
        if (l1>r1 || l2>r2) return nullptr;
        int root_val = postorder[r2];
        int root_idx;
        for (root_idx = 0; inorder[root_idx]!=root_val; root_idx ++);
        TreeNode *root = new TreeNode(root_val);
        root->left = buildTreeWithSegment(inorder, l1, root_idx-1, postorder, l2, l2+root_idx-l1-1);
        root->right = buildTreeWithSegment(inorder, root_idx+1, r1, postorder, l2+root_idx-l1, r2-1);
        return root;
    }
};
